ICU-23222 Enhance unit alias handling

younies opened the pull request:
# Summary

In this PR, we **introduced unit aliases** into the **unit parser Trie**.  When the parser **encounters a unit alias** (e.g., **"permillion" → "part-per-1e6"**),  it **ignores the alias**, **replaces it with the canonical unit identifier**,  and then **continues parsing** as if the alias had never appeared.  

This guarantees that **all units with aliases are normalized** to their replacements,  ensuring **consistent parsing results** across both aliases and canonical forms.

### Example:

```
Input: "meter-permillion-second"
          │
          ▼
[1] Parser reads "meter"
    ✔ Parsed: [meter]
    Remaining: "permillion-second"

[2] Parser reads "permillion"
    🔎 Alias detected → replace with "part-per-1e6"
    ✔ Parsed so far: [meter]
    Remaining becomes: "part-per-1e6-second"

[3] Continue parsing...
    - "part"   → Parsed: [meter, part]
    - "per"    → Parsed: [meter, part, per]
    - "1e6"    → Parsed: [meter, part, per, 1e6]
    - "second" → Parsed: [meter, part, per, 1e6, second]

[✔] Final Result:
    "meter-part-per-1e6-second"
```
#### Before / After (at a glance)

Input : meter-permillion-second
Output: meter-part-per-1e6-second

#### Mini Parse Tree 

```
UnitsExpression
├─ Unit("meter")
├─ Unit("part")
├─ Token("per")
├─ Const("1e6")
└─ Unit("second")
```

-----------------
# Problem Statement

The problem is that we need to scan a CLDR unit ID (string) for any known aliases and replace each alias with its canonical replacement.

## Options

### 1) Brute force

Scan for each alias across the entire string; when an alias is found, replace it in-place (or build a new string with the replacement).

Pros
- Simple to implement.

Cons
- Performance: O(N × M), where N = length of the ID and M = number of aliases (each alias checked against the whole string).


### 2) Dedicated alias Trie

Build a separate Trie containing all aliases. At the start of beside the parsing Trie, run a pass over the CLDR ID to find/replace aliases using this Trie.

Pros
- Completes alias work up-front; parsing code doesn’t need to think about aliases afterward.

Cons
- Extra pass on every parse to check for aliases (often there are none).
- The mutated ID must represent the entire CLDR ID (see Option 3 which does need only the remainig part (unparsed part)).
- Requires maintaining a second Trie alongside the Parsing Trie.


### 3) Integrate aliases into the Parsing Trie (current PR)

Add alias entries directly to the existing Parsing Trie. When the parser encounters an alias token, it ignores that token, splices in the alias’s replacement, and continues parsing the remaining (unparsed) tail of the CLDR ID.

Pros
- No pre-scan for aliases; replacements occur naturally during parsing.
- No need to rewrite/add the entire string—only the unparsed remainder after the alias.
- No separate alias Trie to maintain.

Cons
- Less obvious than Option 2 and requires careful handling of token boundaries and continuation logic.

-----------------
# Implementations

## ICU4J

- Refactor NumberFormatterApiTest to improve test case structure for unit aliases.
- Introduce UnitAliases class to manage unit aliases and their replacements.
- Update UnitsData to load aliases and integrate them into the unit parsing logic.
- Modify LongNameHandler to prepare for future refactoring related to unit aliases.
- Add alias handling in MeasureUnitImpl to support alias replacements during parsing.

## ICU4C

- Introduced a new UnitAliasesSink class to collect unit aliases and their replacements from resource tables.
- Added support for alias handling in the Parser class, allowing for dynamic replacement of unit identifiers.
- Updated the initUnitExtras function to populate global vectors for unit aliases and replacements.
- Enhanced NumberFormatterApiTest to include tests for new unit aliases and ensure compatibility with existing formats.
- Added comments for future refactoring of LongNameHandler to utilize the new aliasing system.

---------------
#### Checklist
- [x] Required: Issue filed: ICU-23222
- [x] Required: The PR title must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

younies commented on icu4j/main/common_tests/src/test/java/com/ibm/icu/dev/test/number/NumberFormatterApiTest.java:
```diff
@@ -6012,29 +6012,65 @@ class TestCase {
-                new TestCase(MeasureUnit.MILLIGRAM_PER_DECILITER, "2 milligrams per deciliter"),
-                new TestCase(MeasureUnit.LITER_PER_100KILOMETERS, "2 liters per 100 kilometers"),
-                new TestCase(MeasureUnit.PART_PER_MILLION, "2 parts per million"),
-                new TestCase(MeasureUnit.MILLIMETER_OF_MERCURY, "2 millimeters of mercury"),
```
It is not taken out—you can see in the code below that all of them are listed. I just wanted to organize them with the other aliases in groups.

younies commented on icu4j/main/core/src/main/java/com/ibm/icu/impl/units/MeasureUnitImpl.java:
```diff
@@ -750,6 +775,7 @@ private SingleUnitOrConstant nextSingleUnit() {
 
             // Read tokens until we have a complete SingleUnit or we reach the end.
             while (true) {
+                Token.Type type = token.getType();
```
done

richgillam commented on icu4c/source/i18n/measunit_extra.cpp:
```diff
@@ -255,6 +258,76 @@ class SimpleUnitIdentifiersSink : public icu::ResourceSink {
+
+        // Collect the unit aliases from the table resource.
+        const char *unitAlias;
+        for (int32_t i = 0; table.getKeyAndValue(i, unitAlias, value); ++i) {
```
You might want to consider using `getAllChildrenWithFallback()` instead of `getAllItemsWithFallback()`.  It'd let you avoid this loop and might help if instances of this resource ever use sideways inheritance.

younies replied:
Great suggestion, done :)

richgillam commented on icu4c/source/i18n/measunit_extra.cpp:
```diff
@@ -255,6 +258,76 @@ class SimpleUnitIdentifiersSink : public icu::ResourceSink {
+        for (int32_t i = 0; table.getKeyAndValue(i, unitAlias, value); ++i) {
+            int32_t unitAliasLen = static_cast<int32_t>(uprv_strlen(unitAlias));
+            CharString *unitAliasCharStr = unitAliases.emplaceBackAndCheckErrorCode(status);
+            if (U_FAILURE(status)) {
```
If you functions you're calling do the right thing, you shouldn't have to call `U_FAILURE()` after every call.  If we've done our job right, each function should check the error code right at the top and dump out without doing anything if it's already showing an error value.  That will usually let you perform a bunch of operations at once and only check the error code at the end, making code like this a lot easier to read.

In this particular case, that'd also allow you to combine these things into one line:

```
unitAliases.emplaceBackAndCheckErrorCode(status)->append(unitAlias, unitAliasLen, status);
```

I personally don't think the `emplaceBack()` thing makes a whole lot of sense (or reads well) when it's split up into two lines like this.  If you keep it as two lines, it seems like you'd be better off just creating a `CharString` on the stack, setting it to the correct value, and then appending it to the `unitAliases` vector.

richgillam commented on icu4c/source/i18n/measunit_extra.cpp:
```diff
@@ -923,6 +1034,36 @@ class Parser {
+                auto replacement = gUnitReplacements[aliasIndex];
+
+                // Create new source string: replacement + remaining unparsed portion
+                fModifiedSource.clear();
```
If I understand correctly, you've incorporated the alias mechanism directly into your parser, but then, if you encounter a token with an alias, you just create a new version of the unit string and copy the alias's replacement text into the string you're parsing and continue parsing the _new_ string instead.  I wonder if it'd make things simpler to handle alias replacement in a separate preprocessing step instead of incorporating it directly into your parser?

(I'm not completely sold on this suggestion-- it may actually be simpler the way you're doing it now.)

richgillam replied:
Yeah, I think you're right.  Thanks for enumerating all this.

richgillam commented on icu4c/source/test/intltest/numbertest_api.cpp:
```diff
@@ -6063,30 +6063,95 @@ void NumberFormatterApiTest::formatUnitsAliases() {
-        {std::make_unique<MeasureUnit>(MeasureUnit::getPartPerMillion()), nullptr, u"2 parts per million"},
-        {std::make_unique<MeasureUnit>(MeasureUnit::getMillimeterOfMercury()), nullptr, u"2 millimeters of mercury"},
-
-        // Some replacements
```
You no longer have a comment over the new entries below saying what they are.  I'm guessing it'd be "part-per-million"?

richgillam reviewed (approved):
I think this looks pretty good.  I'd urge you to consider my suggestions, but I can live with it if this goes in as-is.  Thanks for addressing my previous comments.

younies commented:
> > > How were you handling aliases before, and can you provide a summary explaining what you've changed?
> > > What's the plan for C++? I'm assuming this isn't a Java-only problem.
> > 
> > 
> > I have added a summary for this PR, and I also merged the C++ changes into it instead of creating separate PRs.
> 
> I read your new summary, and it looks good, but it doesn't answer my first question above: How were you handling these aliases before? Is the answer that we _weren't_ handling aliases before now?

We were not handling the aliases at all at the parsing level. we are handling then at the formatting level. I have added a todo there now for removing/modifying this handling after this PR.

richgillam commented on icu4c/source/i18n/measunit_extra.cpp:
```diff
@@ -255,6 +258,58 @@ class SimpleUnitIdentifiersSink : public icu::ResourceSink {
+  private:
+    MaybeStackVector<CharString> &unitAliases;
+    MaybeStackVector<CharString> &unitReplacements;
+};
```
This looks a lot better!  Thanks for making the change.

richgillam reviewed (approved):
Looks good!  Thanks for the changes and explanations.
